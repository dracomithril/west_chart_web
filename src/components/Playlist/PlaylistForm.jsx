/**
 * Created by michal.grezel on 10.04.2017.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSave, faLightbulb } from '@fortawesome/free-solid-svg-icons';
import { weekInfo } from '../../utils/utils';
import './playlist.css';

export default class PlaylistForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      playlistName: '',
      isPrivate: false,
    };
  }

  onGenPlaylistName = () => {
    const { monday, friday } = weekInfo();

    const mondayStr = monday.format('MMM_D').toUpperCase();
    const fridayStr = friday.format('MMM_D').toUpperCase();

    const playlistName = `Chart_${mondayStr}-${fridayStr}`;
    this.setState({ playlistName });
  };

  render() {
    const { store } = this.context;
    const { spotifyUser } = store.getState();
    const {
      onCreatePlaylistClick: onCreatePlaylist, onStartClick, hasElements, isUserLogged,
    } = this.props;
    const { playlistName, isPrivate } = this.state;
    const disable_create = !(playlistName.length > 5 && hasElements && isUserLogged);
    const validatePlaylistName = (str = '') => str.length > 7;
    const isNameValid = validatePlaylistName(playlistName);
    return (
      <div className="playlist-form">
        <Button variant="raised" color="primary" onClick={onStartClick} id="start_sp_button">
          start
        </Button>
        <TextField
          error={!isNameValid}
          value={playlistName}
          style={{ padding: '5px' }}
          placeholder="playlist name"
          onChange={({ target }) => {
            this.setState({ playlistName: target.value });
          }}
        />
        <div style={{ alignSelf: 'center' }}>
          <Button variant="fab" color="secondary" onClick={this.onGenPlaylistName} id="genName_sp_button">
            <FontAwesomeIcon icon={faLightbulb} />
          </Button>
          <Button
            variant="fab"
            id="crt_pl_button"
            onClick={() => onCreatePlaylist && onCreatePlaylist({ playlistName, isPrivate })}
            disabled={!isNameValid && Boolean(spotifyUser.id)}
            title={disable_create ? 'Sorry you need to be logged to spotify to be able to add playlist' : ''}
          >
            <FontAwesomeIcon icon={faSave} />
          </Button>
          <FormControlLabel
            control={(
              <Checkbox
                id="play_list_is_private"
                value="private"
                color="primary"
                checked={isPrivate}
                onChange={({ target }) => {
                  this.setState({ isPrivate: target.checked });
                }}
              />
)}
            label="private"
          />
        </div>
      </div>
    );
  }
}
PlaylistForm.contextTypes = {
  store: PropTypes.shape(),
};
PlaylistForm.propTypes = {
  hasElements: PropTypes.bool,
  isUserLogged: PropTypes.bool,
  onStartClick: PropTypes.func,
  onCreatePlaylistClick: PropTypes.func,
};
