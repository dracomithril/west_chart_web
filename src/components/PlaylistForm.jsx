/**
 * Created by Gryzli on 10.04.2017.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonGroup, Form, FormControl, FormGroup, InputGroup } from 'react-bootstrap';
import { searchForMusic, createPlaylistAndAddTracks } from '../utils/spotify_utils';
import { getArtist_Title, weekInfo } from '../utils/utils';

import action_types from './../reducers/action_types';

export default class PlaylistForm extends Component {
  /* istanbul ignore next */
  componentDidMount() {
    console.info('component PlaylistForm did mount');
  }

  /* istanbul ignore next */
  componentWillUnmount() {
    console.info('component PlaylistForm unmounted');
  }

  onStartClick = () => {
    const { store } = this.context;
    const { selected } = this.props;
    const search = selected.map((elem, search_id) => {
      const entry = getArtist_Title(elem.link.title);
      const search_track = {
        artist: entry.artist,
        title: entry.title,
        full_title: elem.link.title,
        id: elem.id,
        search_id,
        items: [],
        selected: {},
      };
      searchForMusic(search_track, store).then(res =>
        store.dispatch({
          type: action_types.UPDATE_SINGLE_SEARCH,
          field: 'items',
          value: res.value,
          id: res.id,
        }),
      );
      return search_track;
    });
    store.dispatch({ type: action_types.UPDATE_SEARCH, search });
  };

  onGenPlaylistName = () => {
    const { store } = this.context;
    const { monday, friday } = weekInfo(new Date());

    const monday_str = monday
      .toLocaleString('en-US', { month: 'short', day: 'numeric' })
      .toUpperCase();
    const friday_str = friday
      .toLocaleString('en-US', { month: 'short', day: 'numeric' })
      .toUpperCase();

    const playlist_name = `Chart_${monday_str}-${friday_str}`;
    const list = playlist_name.split(' ').join('_');
    store.dispatch({ type: action_types.UPDATE_PLAYLIST_NAME, value: list });
  };

  onCreatePlaylist = () => {
    const { store } = this.context;
    const { search_list, sp_user, sp_playlist_name, isPlaylistPrivate } = store.getState();
    const selected = search_list
      .map(elem => (elem.selected !== undefined ? elem.selected.uri : undefined))
      .filter(elem => elem !== undefined);
    createPlaylistAndAddTracks(sp_user, sp_playlist_name, isPlaylistPrivate, selected).then(info =>
      store.dispatch({
        type: action_types.UPDATE_PLAYLIST_INFO,
        value: info,
      }),
    );
  };

  render() {
    const { store } = this.context;
    const { sp_playlist_name, isPlaylistPrivate } = store.getState();
    const { selected } = this.props;
    const disable_create = !(sp_playlist_name.length > 5 && selected.length > 0);
    const validatePlaylistName = strLength => {
      if (strLength > 8) {
        return 'success';
      }
      return strLength > 5 ? 'warning' : 'error';
    };
    return (
      <Form inline>
        <Button onClick={this.onStartClick} id="start_sp_button" bsStyle="success">
          Start
        </Button>
        <FormGroup
          style={{ margin: '1px 5px 5px 5px' }}
          controlId="play_list_name"
          validationState={validatePlaylistName(sp_playlist_name.length)}
        >
          <InputGroup style={{ maxWidth: 250 }}>
            <FormControl
              type="text"
              placeholder="playlist name"
              value={sp_playlist_name}
              onChange={e => {
                store.dispatch({
                  type: action_types.UPDATE_PLAYLIST_NAME,
                  value: e.target.value,
                });
              }}
            />
            <FormControl.Feedback />
            {/* <InputGroup.Addon><Glyphicon glyph="music"/></InputGroup.Addon> */}
          </InputGroup>
        </FormGroup>
        <ButtonGroup>
          <Button onClick={this.onGenPlaylistName} id="genName_sp_button" bsStyle="primary">
            gen. name
          </Button>
          <Button
            id="crt_pl_button"
            onClick={this.onCreatePlaylist}
            disabled={disable_create}
            bsStyle="danger"
          >
            <i className="fas fa-save" /> Playlist
          </Button>
        </ButtonGroup>
        <label htmlFor="play_list_is_private">
          <input
            type="checkbox"
            id="play_list_is_private"
            onChange={e => {
              store.dispatch({
                type: action_types.TOGGLE_IS_PRIVATE,
                value: e.target.checked,
              });
            }}
            value="private"
            checked={isPlaylistPrivate}
          />
          {'private ?'}
        </label>
      </Form>
    );
  }
}
PlaylistForm.contextTypes = {
  store: PropTypes.object,
};
PlaylistForm.propTypes = {
  selected: PropTypes.array,
};
