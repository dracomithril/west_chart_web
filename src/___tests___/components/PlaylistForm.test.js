/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import PlaylistForm from './../../components/PlaylistForm';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

jest.mock('./../../utils/spotify_utils');
Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../data/initial_state.json');
const data = require('../data/response.json').chart;
const sinon = require('sinon');
const mockStore = configureMockStore([]);

describe('<PlaylistForm/>', () => {
  it('renders without crashing', () => {
    const store = mockStore(initial_state);
    const wrapper = shallow(
      <PlaylistForm/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('start_click', (done) => {
    const sp_utils = require('../../utils/spotify_utils');
    sp_utils.searchForMusic.mockReturnValue(Promise.resolve({ id: 'zzz', value: {} }));
    const state = Object.assign({}, initial_state, { sp_playlist_name: "test_list_zzzz" });
    const store = mockStore(state);
    const wrapper = shallow(
      <PlaylistForm selected={data}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    let start_b = wrapper.find('#start_sp_button');
    start_b.simulate('click');
    expect(store.getActions().length).toBe(1);
    done();
  });
  it('createPlaylistAndAddTracks', () => {
    const sp_utils = require('../../utils/spotify_utils');
    sp_utils.createPlaylistAndAddTracks.mockReturnValue(Promise.resolve({ mock: 'mock' }));
    const state = Object.assign({}, initial_state, { sp_playlist_name: "test_list_zzzz" });
    const store = mockStore(state);
    const wrapper = shallow(
      <PlaylistForm selected={data}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );

    let elem = wrapper.find('#play_list_is_private');
    let getElemByIdStub = sinon.stub();
    getElemByIdStub.withArgs('play_list_is_private').returns(elem);
    //document.getElementById = getElemByIdStub;
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    let start_b = wrapper.find('#crt_pl_button');
    start_b.simulate('click');
    expect(store.getActions().length).toBe(0);
    expect(sp_utils.createPlaylistAndAddTracks.mock.calls.length).toBe(1);
  })
});
