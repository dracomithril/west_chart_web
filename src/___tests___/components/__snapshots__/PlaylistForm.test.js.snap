// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistForm/> createPlaylistAndAddTracks 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "margin": "1px 5px 5px 5px",
      }
    }
    validationState="success"
  >
    <InputGroup
      bsClass="input-group"
      style={
        Object {
          "maxWidth": 250,
        }
      }
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value="test_list_zzzz"
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <ButtonGroup
    block={false}
    bsClass="btn-group"
    justified={false}
    vertical={false}
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      id="genName_sp_button"
      onClick={[Function]}
    >
      gen. name
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="danger"
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
    >
      <i
        className="fa fa-save"
      />
       Playlist
    </Button>
  </ButtonGroup>
  <label
    htmlFor="play_list_is_private"
  >
    <input
      checked={false}
      id="play_list_is_private"
      onChange={[Function]}
      type="checkbox"
      value="private"
    />
    private ?
  </label>
</Form>
`;

exports[`<PlaylistForm/> renders without crashing 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "margin": "1px 5px 5px 5px",
      }
    }
    validationState="error"
  >
    <InputGroup
      bsClass="input-group"
      style={
        Object {
          "maxWidth": 250,
        }
      }
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value=""
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <ButtonGroup
    block={false}
    bsClass="btn-group"
    justified={false}
    vertical={false}
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      id="genName_sp_button"
      onClick={[Function]}
    >
      gen. name
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="danger"
      disabled={true}
      id="crt_pl_button"
      onClick={[Function]}
    >
      <i
        className="fa fa-save"
      />
       Playlist
    </Button>
  </ButtonGroup>
  <label
    htmlFor="play_list_is_private"
  >
    <input
      checked={false}
      id="play_list_is_private"
      onChange={[Function]}
      type="checkbox"
      value="private"
    />
    private ?
  </label>
</Form>
`;

exports[`<PlaylistForm/> start_click 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="success"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "margin": "1px 5px 5px 5px",
      }
    }
    validationState="success"
  >
    <InputGroup
      bsClass="input-group"
      style={
        Object {
          "maxWidth": 250,
        }
      }
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value="test_list_zzzz"
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <ButtonGroup
    block={false}
    bsClass="btn-group"
    justified={false}
    vertical={false}
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      id="genName_sp_button"
      onClick={[Function]}
    >
      gen. name
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="danger"
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
    >
      <i
        className="fa fa-save"
      />
       Playlist
    </Button>
  </ButtonGroup>
  <label
    htmlFor="play_list_is_private"
  >
    <input
      checked={false}
      id="play_list_is_private"
      onChange={[Function]}
      type="checkbox"
      value="private"
    />
    private ?
  </label>
</Form>
`;
